### ISSUE TITLE
Discussion: Relationships support

### ISSUE THREAD

**Issue by wovalle** [wovalle/fireorm#32]
I'll start the design process for the Relationships feature. So far the idea is to express explicit relationships between two collection using the provided decorators.

@OneToOne
@OneToMany
@ManyToMany

Also, the relationships should provide a way to declare them as Lazy or Eager (and Lazy will be the default). Fireorm will retrieve all the eager relationships when doing a query and will return (non-evaluated) Promises for the lazy ones.

Open Questions:
- Should we support @ManyToOne?
- The whole @ManyToMany process. In traditional SQL that's achieved by creating an intermediate table which fireorm shouldn't be responsible of.
- Since Promises are eagerly evaluated, how the Lazy Promise implementation would look like.
- Decorators signature

**Comment by jonesnc** [comment]
I'll start thinking about this as well. My initial thoughts are regarding this question:

> Promises are eagerly evaluated, how the Lazy Promise implementation would look like.

A possible solution to this would be to create a function that returns the "Lazy Promise". This will ensure that the "Lazy Promise" will only begin executing when we call the function that returns the "Lazy Promise".

Something like:

```
function createLazyPromise() {
    const lazyPromise = new Promise((resolve, reject) => {
        // execute query
    });
    return lazyPromise;
}
// lazyPromise has not begun "executing" here, since the `createLazyPromise` function hasn't been called yet.
```

**Comment by wovalle** [comment]
Yeah I thought something similar, but the problem is that the `relation field` will be initialize as a function, but then it'll be updated with the actual data, and I don't want that. 

Given this entity:

```typescript
export class Band {
  id: string;
  name: string;
  formationYear: number;
  genres: Array<string>;

  @OneToMany(Album, (album, band) => a.bandId === band.id, {lazy: true})
  albums: ISubCollection<Album>;
}

const band = await bandRepository.findById('three-days-grace');
// Calling band.albums to retrieve the albums
await band.albums();

// Now albums is not a function anymore
console.log(band.albums)
```

Note: that syntaxis is not final, haven't had time to actually design how it would work

**Comment by wovalle** [comment]
Started with the actual implementation, a lot of things are missing.
